datasets/dataset_esc50.py

```diff
@@ -0,0 +1,158 @@
+# -*- coding: utf-8 -*-
+"""
+ESC-50 dataset
+More infos at https://github.com/karolpiczak/ESC-50
+"""
+import glob
+import os
+import shutil
+from urllib.request import urlretrieve
+import zipfile
+from pathlib import Path
+
+import pandas as pd
+import numpy as np
+
+from librosa import load as wav_load
+from librosa.feature import mfcc
+from sklearn.model_selection import train_test_split
+from loguru import logger
+
+from keras.models import Sequential
+from keras.layers import Dense, Dropout
+from keras.layers import Conv2D, GlobalAveragePooling2D, MaxPooling2D
+from keras.losses import categorical_crossentropy
+from keras.utils import to_categorical
+
+from . import dataset
+
+num_classes = 50
+input_shape = (40, 431, 1)
+
+
+# return a dataset object
+def generate_new_dataset():
+    # Load data
+    (x_train, y_train), (x_test, y_test) = load_data()
+
+    # Pre-process inputs
+    logger.info('Preprocessing the raw audios')
+    x_train = preprocess_dataset_inputs(x_train)
+    x_test = preprocess_dataset_inputs(x_test)
+
+    dataset_obj = dataset.Dataset(
+            "esc50",
+            x_train,
+            x_test,
+            y_train,
+            y_test,
+            input_shape,
+            num_classes,
+            preprocess_dataset_labels,
+            generate_new_model_for_dataset,
+        )
+    return dataset_obj
+
+# Init dataset-specific functions
+def preprocess_dataset_labels(y):
+    num_classes = 50
+    y = to_categorical(y, num_classes)
+
+    return y
+
+
+def preprocess_dataset_inputs(x):
+    """
+    Compute the Mel-Frequency Cepstral Coefficients
+    :param x: iterator. Yield tuples, of audio and rate.
+    :return: Array of mfcc images.
+    """
+
+    features_list = []
+    for audio, rate in x:
+        mfccs = mfcc(y=audio, sr=rate, n_mfcc=40)
+        # mfccs_scaled = np.mean(mfccs.T, axis=0)
+        features_list.append(mfccs)
+    features = np.array(features_list)
+    return features.reshape(((features.shape[0],) + input_shape))
+
+
+# download train and test sets
+def load_data():
+    """
+    load the dataset. Note that the x are iterators which need to be preprocess
+    :return: (x_train, y_train), (x_test, y_test)
+    """
+    path = Path(__file__).resolve().parents[0]
+    repertoire = str(path) + '/local_data/esc50'
+    if not os.path.exists(repertoire):
+        os.makedirs(repertoire)
+        os.chdir(repertoire)
```

Comment:
[Pathlib ](https://docs.python.org/fr/3/library/pathlib.html)is a great alternative to `os.path` !